src/ray/common/ray_object.cc

```diff
@@ -96,12 +95,7 @@ RayObject::RayObject(rpc::ErrorType error_type, const rpc::RayErrorInfo *ray_err
     return;
   }
 
-  RAY_CHECK(ray_error_info->has_actor_init_failure());
-  // This is temporarily here because changing this requires changes in all language
-  // frontend.
-  // TODO(sang, lixin): Remove it.
-  const auto error_buffer =
-      MakeSerializeErrorBuffer<rpc::RayException>(ray_error_info->actor_init_failure());
+  const auto error_buffer = MakeSerializedErrorBuffer<rpc::RayErrorInfo>(*ray_error_info);
```

Comment:
@lixin-wei can you help fixing the Java case? Java side will be incompatible to this change.

Reply:
The code has changed a lot since my last modification. I read your discussion but didn't get the context. Would you please say more about what happened, why it will affect java cases, and what should I do?

Reply:
Previously, all of ray error object didn't have body, except the actor creation exception. Now, we want to allow to pass protobuf-defined body so that we can write a better error message on failures. 

Currently, we allow to specify body for all ray objects. And by default we changed this to `message RayErrorInfo `, and `RayErrorInfo ` contains `RayException` (which was for actor creation exception). This is incompatible to Java / Cpp (since they assume if there's any body on ray object, it is `RayException`, but it is `RayErrorInfo` now).

Reply:
You can see `serialization.py`

Reply:
I see, thanks! But before that, I've some questions, commented.

Reply:
Thanks! I have no Ray java setup on my machine haha..

Reply:
Do you know who I should contact to for cpp API? I assume Cpp APi also needs some changes

Reply:
(though for some reasons the cpp tests don't seem to fail!)

Reply:
Since this PR is large enough, I think maybe it's better to keep the compatibility in this PR, and open follow-up PRs to change Java/CPP deserialization. Do you think so? Then you won't be blocked by us.

You can write the `RayException` in `RayError. actor_died_error` to object store first to keep the compatibility.

Reply:
For CPP, maybe you can ping @qicosmos

Reply:
> Since this PR is large enough, I think maybe it's better to keep the compatibility in this PR, and open follow-up PRs to change Python/Java/CPP deserialization

What do you mean by "keep compatibility" here? I am asking it because this is the PR that breaks the compatibility (all other PRs are already merged).

Reply:
What part of this PR do you think can be isolated in this case?

Reply:
I meant writing RayException to object store first(But keep all refactor you made). And change it to write RayError in the next PR.

Anyway, I'm fine with both approaches. I think 50+ lines will be modified.

I asked @qicosmos, he said CPP worker doesn't support exception now. So you we can leave CPP worker as it is, I think.

Reply:
Hmm I see. So your suggestion is to only modify cpp side in this PR, and make another PR to actually change it to RayErrorInfo.

I am fine with that approach. Lmk what you prefer.

Reply:
I prefer opening another PR. I'll open a PR to change both Python and Java after you merge this PR(for Python I'll copy the code you wrote in this PR). Do you think it's OK?

Reply:
Oh nope, I think I prefer not opening another PR now. Since I realized that I need to copy and debug your python test.

I'll fix the Java test in this PR!

Reply:
Sounds good haha